rfc/rfc-65/rfc-65.md

```diff
@@ -0,0 +1,163 @@
+
+We need persistent partition ttt policies in hudi table config and users should interact with hudi only by
+setting/getting ttl policies. Hudi then will takes care of doing the ttl management in an async table service.
+
+The partition TTL policies will be stored in `hoodie.properties` and they are defined as follows. Explicit policies are
+defined using a JSON array, note we support regex here.
+
+```properties
+# TTL policy definition
+hoodie.partition.ttl.policies=[{"partitionSpec"\:"user_id\=*/","policy"\:"KEEP_BY_TIME","policyValue"\:7},{"partitionSpec"\:"user_id\=3/","policy"\:"KEEP_BY_TIME","policyValue"\:30}]
+```
+
+### Partition Statistics
+
+We need to gather proper partition statistics for different kinds of partition ttl policies. Specifically, for
+KEEP_BY_TIME policy that we want to support first, we need to know the last modified time for every partition in hudi
+table.
+
+To simplify the design, we decide to use the largest commit time of committed file groups in the partition as partition
+last modified time. But for file groups generated by replace commit, it may not reveal the real insert/update time for
+the file group. We can also introduce other mechanism to get last modified time of the partition, like add a new kind of
+partition metadata or add a new field in metadata table. Open to ideas for this design choice.
+
+### Executing TTL management
+
+Once we have defined the TTL policies and have proper partition statistics, it's easy to apply the policies and delete
+expired partitions. The process of TTL management is very similar to other table services in hudi, we will provide a new
+method called `managePartitionTTL` in `BaseHoodieTableServiceClient` and we can invoke this method in async table
+service, SparkSQL Call Command, cli, JAVA code etc.
+
+We will provide an Async Table Service as the default interface to the users that want to do partition TTL management
+and we can add more interfaces in the future.
+
+User will config the async ttl management service as follows. TTL management will be trigger N commits after last
+TTL replace commit.
+
+```properties
+hoodie.partition.ttl.management.async=true
+hoodie.partition.ttl.management.min.commits=10
+```
+
+The process of manage partition TTL is as follows:
+
+1. Gather partitions statistics, list all partitions and the largest commit time of committed
+   file groups of each partition as their last modified time.
+2. Apply each TTL policy. For explicit policies, find sub-partitions matched `PartitionSpec` defined in the policy and
+   check if there are expired partitions according to the policy type and size. For default policy, find partitions that
+   do not match any explicit policy and check if they are expired.
+3. For all expired partitions, call delete_partition to delete them, which will generate a replace commit and mark all
+   files in those partitions as replaced. For pending clustering and compaction that affect the target partition to
```

Comment:
If there is pending clustering and compaction, the ttl policy execution should clean and delete the expired file groups after clustering and compaction. The user doesn't need do any behavior for pending clustering and compaction case.